File: demo-2017-10-08-android/Demo/core/src/main/java/com/example/interactors/createcat/CreateCatInteractor.java
package com.example.interactors.createcat;

import com.example.entities.Cat;
import com.example.repositories.CatsRepository;

import java.util.UUID;

/**
 * Created by Dev on 08/10/2017.
 */

public class CreateCatInteractor {

    private final CatCreatedPresenter presenter;
    private final CatsRepository repository;
    private final CreateCatValidator validator;

    public CreateCatInteractor(
            CatCreatedPresenter presenter,
            CatsRepository repository,
            CreateCatValidator validator) {
        this.presenter = presenter;
        this.repository = repository;
        this.validator = validator;
    }

    public void execute(CreateCatRequest request) {
         CreateCatValidator.ValidationResult validationResult =
                 validator.validate(request);

        if (!validationResult.isValid()) {
            presenter.showValidationError(validationResult.errorMessage);
            return;
        }

        Cat cat = new Cat(UUID.randomUUID(), request.name);
        cat.setTitle(request.title);
        repository.save(cat);

        CreateCatResponse response = new CreateCatResponse();
        response.id = cat.getId();
        response.grandioseName = cat.grandioseName();
        response.isImportant = cat.isImportant();

        presenter.present(response);
    }
}

File: 2017-10-DevconDavao/CatsDemo_OutsideInTDD_DotNetCore/src/Interactors/CreateCat/CreateCatInteractor.cs
using Domain.Entities;
using Domain.Repositories;
using System;

namespace Interactors.CreateCat
{
    public class CreateCatInteractor : ICreateCatInteractor
    {
        private ICreateCatPresenter presenter;
        private ICatsRepository repository;
        
        public CreateCatInteractor(ICreateCatPresenter presenter, ICatsRepository repository)
        {
            this.presenter = presenter;
            this.repository = repository;
        }

        public void Execute(CreateCatRequest request)
        {
            Cat cat = Cat.NewCat();
            cat.Name = request.Name;
            cat.Title = request.Title;

            repository.Save(cat);

            var response = new CreateCatResponse(cat.Id, request.Name, request.Title);
            presenter.Execute(response);
        }
    }
}
